File: ed.py

def distribute_list_continuously(lst, num_groups):
    """Distribute elements of lst into num_groups continuously."""
    if num_groups <= 0:
        raise ValueError("Number of groups must be positive.")

    # Calculate the size of each group
    group_size = len(lst) // num_groups
    remainder = len(lst) % num_groups

    groups = []
    start_index = 0

    for i in range(num_groups):
        # Determine the size of the current group
        # Add 1 to the first 'remainder' groups to distribute the remainder items
        current_group_size = group_size + (1 if i < remainder else 0)
        
        # Slice the list to get the current group
        group = lst[start_index:start_index + current_group_size]
        groups.append(group)
        
        # Update the starting index for the next group
        start_index += current_group_size

    return groups

# Example usage
original_list = ['C', '?', 'T', 'f', 'o', 'V', '4', 'W', 'S', 'n', 'w', '>', 'Y', '9', 'X', 'z', '&', '.', 'd', ')', 'x', '2', '$', 'L', '[', 'Z', '`', '<', 'E', 'e', '~', 'i', 'I', 'k', 'j', '=', ':', 'A', 'b', '_', 'z#x#x', 'Q', '5', 'g', '7', '1', 'l', '+', ']', '{', 'c', '(', 'D', 'r', 'N', '|', ';', '8', '@', 'v', '\\', 'G', '-', '"', 'H', 'O', '}', 'p', ',', 'y', "'", '6', '^', '*', 'B', 'K', 'P', 'R', 'M', 's', 'a', '%', '3', 'u', 'U', 't', 'q', '/', 'J', '!', 'h', 'F', 'm', '0', '#', 'z#y#y']
num_groups = 12
distributed_groups = distribute_list_continuously(original_list, num_groups)


def derangements(lst):
    """Generate all derangements of the list."""
    if len(lst) <= 1:
        return []

    # Helper function to generate derangements recursively
    def generate_derangements(current, remaining):
        if not remaining:
            # If nothing remains, we have a valid derangement
            results.append(current)
            return
        
        for i in range(len(remaining)):
            # Choose an element from remaining
            next_elem = remaining[i]
            # Check if the chosen element is not at its original position
            if next_elem != lst[len(current)]:
                # Create new lists for next recursion
                new_current = current + [next_elem]
                new_remaining = remaining[:i] + remaining[i+1:]
                # Recurse with updated lists
                generate_derangements(new_current, new_remaining)

    results = []
    generate_derangements([], lst)
    return results

all_derangements = [[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]]
for i in range(12):
    all_derangements[i] = derangements(distributed_groups[i])

b = [[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]]
def split_48_digit_integer():
    # Prompt for a 48-digit integer input
    input_integer = input("Enter 48 digits of secret code : ")

    # Store each 4-digit segment in a list
    segments = [input_integer[i:i+4] for i in range(0, 48, 4)]

    # Unpack the list into separate variables
    b[0], b[1], b[2], b[3], b[4], b[5], b[6], b[7], b[8], b[9], b[10], b[11] = segments

    # Print the stored values
    for index, value in enumerate(segments, start=1):
        b[index-1] = int(value)

# Example usage






def searchr(matrix, target):
    """Search for a target value in a 2D list (matrix).
    
    Returns the position (row, column) of the target if found,
    otherwise returns None.
    """
    for row_index in range(len(matrix)):
        for col_index in range(len(matrix[row_index])):
            if matrix[row_index][col_index] == target:
                return row_index
    return None
def searchc(matrix, target):
    """Search for a target value in a 2D list (matrix).
    
    Returns the position (row, column) of the target if found,
    otherwise returns None.
    """
    for row_index in range(len(matrix)):
        for col_index in range(len(matrix[row_index])):
            if matrix[row_index][col_index] == target:
                return col_index
    return None


def run(choice):
    if(choice==1):
        msg = input("Enter message : ")
        split_48_digit_integer()
        print("Secret Code : ", end="")
        k = "null"
        for i in range(len(msg)):
            if msg[i] == " ":
                k = "z#y#y"
                print(all_derangements[searchr(distributed_groups, k)][b[searchr(distributed_groups, k)]][searchc(distributed_groups, k)], end="")
            else:
                print(all_derangements[searchr(distributed_groups, msg[i])][b[searchr(distributed_groups, msg[i])]][searchc(distributed_groups, msg[i])], end="")
    elif(choice==2):
        def search(matrix, target):
            """Search for a target value in a 2D list (matrix).
            
            Returns the position (row, column) of the target if found,
            otherwise returns None.
            """
            for index in range(len(matrix)):
                if matrix[index] == target:
                    return index
            return None

        def printf(txt) :
            if(txt=="z#y#y") :
                print(" ", end="")
            else :
                print(txt, end="")
            
        def error() :
            print("\n\033[F\033[K", end='')
            print("You write the wrong code or message....")


        msg = input("Enter message : ")
        split_48_digit_integer()
        print("Secret Code : ", end="")
        k = "null"
        i=0
        j=len(msg)-1
        while i<=j:
            if((msg[i] == "z" and msg[i+1]== "#" and msg[i+3]=="#") and ((msg[i+2]=="y" or msg[i+2]=="x")and((msg[i+4]=="y" or msg[i+4]=="x")))):
                if(msg[i+2]=="y"):
                    try :
                        printf(distributed_groups[searchr(distributed_groups, "z#y#y")][search(all_derangements[searchr(distributed_groups, "z#y#y")][b[searchr(distributed_groups, msg[i])]],"z#y#y")])
                    except TypeError:
                        error()
                        break
                    i += 4
                else :
                    try :
                        printf(distributed_groups[searchr(distributed_groups, "z#x#x")][search(all_derangements[searchr(distributed_groups, "z#x#x")][b[searchr(distributed_groups, msg[i])]],"z#x#x")])
                    except TypeError:
                        error()
                        break
                    i+=4
            else:
                try :
                    printf(distributed_groups[searchr(distributed_groups, msg[i])][search(all_derangements[searchr(distributed_groups, msg[i])][b[searchr(distributed_groups, msg[i])]],msg[i])])
                except TypeError:
                    error()
                    break
            i+=1
    else:
        start()


def start():
    print("1. Encrypt")
    print("2. Decrypt")
    inp = int(input("Enter your choice number : "))
    choice = inp
    run(choice)


start()

File: ed.js
function distributeListContinuously(lst, numGroups) {
  if (numGroups <= 0) {
      throw new Error("Number of groups must be positive.");
  }

  const groupSize = Math.floor(lst.length / numGroups);
  const remainder = lst.length % numGroups;
  const groups = [];
  let startIndex = 0;

  for (let i = 0; i < numGroups; i++) {
      const currentGroupSize = groupSize + (i < remainder ? 1 : 0);
      const group = lst.slice(startIndex, startIndex + currentGroupSize);
      groups.push(group);
      startIndex += currentGroupSize;
  }

  return groups;
}

function derangements(lst) {
  if (lst.length <= 1) return [];

  const results = [];

  function generate(current, remaining) {
      if (remaining.length === 0) {
          results.push(current);
          return;
      }

      for (let i = 0; i < remaining.length; i++) {
          const nextElem = remaining[i];
          if (nextElem !== lst[current.length]) {
              const newCurrent = current.concat([nextElem]);
              const newRemaining = remaining.slice(0, i).concat(remaining.slice(i + 1));
              generate(newCurrent, newRemaining);
          }
      }
  }

  generate([], lst);
  return results;
}

const originalList = ['C', '?', 'T', 'f', 'o', 'V', '4', 'W', 'S', 'n', 'w', '>', 'Y', '9', 'X', 'z', '&', '.', 'd', ')', 'x', '2', '$', 'L', '[', 'Z', '`', '<', 'E', 'e', '~', 'i', 'I', 'k', 'j', '=', ':', 'A', 'b', '_', 'z#x#x', 'Q', '5', 'g', '7', '1', 'l', '+', ']', '{', 'c', '(', 'D', 'r', 'N', '|', ';', '8', '@', 'v', '\\', 'G', '-', '"', 'H', 'O', '}', 'p', ',', 'y', "'", '6', '^', '*', 'B', 'K', 'P', 'R', 'M', 's', 'a', '%', '3', 'u', 'U', 't', 'q', '/', 'J', '!', 'h', 'F', 'm', '0', '#', 'z#y#y'];
const numGroups = 12;
const distributedGroups = distributeListContinuously(originalList, numGroups);

const allDerangements = Array(16).fill().map((_, i) => derangements(distributedGroups[i] || []));

const b = Array(16).fill(0);

function split48DigitInteger(input_integer) {
  const segments = [];
  for (let i = 0; i < 48; i += 4) {
      segments.push(parseInt(input_integer.slice(i, i + 4)));
  }

  for (let i = 0; i < 12; i++) {
      b[i] = segments[i];
  }
}

function searchr(matrix, target) {
  for (let rowIndex = 0; rowIndex < matrix.length; rowIndex++) {
      if (matrix[rowIndex].includes(target)) {
          return rowIndex;
      }
  }
  return null;
}

function searchc(matrix, target) {
  for (let rowIndex = 0; rowIndex < matrix.length; rowIndex++) {
      const colIndex = matrix[rowIndex].indexOf(target);
      if (colIndex !== -1) {
          return colIndex;
      }
  }
  return null;
}

function run(choice) {
  const prompt = require('prompt-sync')();

  if (choice === 1) {
      const msg = prompt("Enter message : ");
      const input_integer = prompt("Enter 48 digits of secret code : ");
      split48DigitInteger(input_integer);

      let output = "Secret Code : ";
      for (let i = 0; i < msg.length; i++) {
          const char = msg[i];
          if (char === " ") {
              const k = "z#y#y";
              const row = searchr(distributedGroups, k);
              const col = searchc(distributedGroups, k);
              output += allDerangements[row][b[row]][col];
          } else {
              const row = searchr(distributedGroups, char);
              const col = searchc(distributedGroups, char);
              output += allDerangements[row][b[row]][col];
          }
      }
      console.log(output);

  } else if (choice === 2) {
      const msg = prompt("Enter message : ");
      const input_integer = prompt("Enter 48 digits of secret code : ");
      split48DigitInteger(input_integer);

      console.log("Secret Code : ");
      let i = 0;
      const j = msg.length - 1;

      function search(arr, val) {
          return arr.indexOf(val);
      }

      function printf(txt) {
          process.stdout.write(txt === "z#y#y" ? " " : txt);
      }

      function error() {
          console.log("You wrote the wrong code or message...");
      }

      while (i <= j) {
          if ((msg[i] === "z" && msg[i + 1] === "#" && msg[i + 3] === "#") &&
              ((msg[i + 2] === "y" || msg[i + 2] === "x") && (msg[i + 4] === "y" || msg[i + 4] === "x"))) {
              if (msg[i + 2] === "y") {
                  try {
                      printf(distributedGroups[searchr(distributedGroups, "z#y#y")][
                          search(allDerangements[searchr(distributedGroups, "z#y#y")][b[searchr(distributedGroups, msg[i])]], "z#y#y")
                      ]);
                  } catch (e) {
                      error(); break;
                  }
                  i += 4;
              } else {
                  try {
                      printf(distributedGroups[searchr(distributedGroups, "z#x#x")][
                          search(allDerangements[searchr(distributedGroups, "z#x#x")][b[searchr(distributedGroups, msg[i])]], "z#x#x")
                      ]);
                  } catch (e) {
                      error(); break;
                  }
                  i += 4;
              }
          } else {
              try {
                  printf(distributedGroups[searchr(distributedGroups, msg[i])][
                      search(allDerangements[searchr(distributedGroups, msg[i])][b[searchr(distributedGroups, msg[i])]], msg[i])
                  ]);
              } catch (e) {
                  error(); break;
              }
          }
          i++;
      }
  } else {
      start();
  }
}

function start() {
  const prompt = require('prompt-sync')();
  console.log("1. Encrypt");
  console.log("2. Decrypt");
  const choice = parseInt(prompt("Enter your choice number : "));
  run(choice);
}

start();
